File: src/main/java/io/github/saneea/feature/binary/gzip/ToGzip.java
package io.github.saneea.feature.binary.gzip;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.zip.GZIPOutputStream;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class ToGzip implements//
		Feature, //
		Feature.In.Bin.Stream, //
		Feature.Out.Bin.Stream {

	private InputStream in;
	private OutputStream out;

	@Override
	public Meta meta(FeatureContext context) {
		return Meta.from("compress to GZIP");
	}

	@Override
	public void run(FeatureContext context) throws IOException {
		try (GZIPOutputStream gzipOut = new GZIPOutputStream(out)) {
			in.transferTo(gzipOut);
		}
	}

	@Override
	public void setIn(InputStream in) {
		this.in = in;
	}

	@Override
	public void setOut(OutputStream out) {
		this.out = out;
	}
}

File: src/main/java/io/github/saneea/dvh/feature/binary/gzip/ToGzip.kt
package io.github.saneea.dvh.feature.binary.gzip

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.FeatureContext
import java.io.InputStream
import java.io.OutputStream
import java.util.zip.GZIPOutputStream

class ToGzip :
    Feature,
    Feature.In.Bin.Stream,
    Feature.Out.Bin.Stream {

    override lateinit var context: FeatureContext
    override lateinit var inBinStream: InputStream
    override lateinit var outBinStream: OutputStream

    override val meta = Meta("compress to GZIP")

    override fun run() {
        GZIPOutputStream(outBinStream).use(inBinStream::transferTo)
    }
}